pkg/iac/terraform/state/backend/tf_config_test.go

```diff
@@ -0,0 +1,88 @@
+package backend
+
+import (
+	"fmt"
+	"io/fs"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+	"testing"
+
+	"github.com/mitchellh/go-homedir"
+)
+
+func TestGetTFCloudToken(t *testing.T) {
+	originalEnv := os.Getenv("HOME")
+	os.Setenv("HOME", "/tmp")
```

Comment:
We don't care about setting a read HOME env var and testing that we get the right one, this is related to the mitchel's `homedir` library. 

Here we want to test two things :
- We want to ensure that the right path is used to read the file
- We want to test the output of the `GetToken` with multiples kind of files

So we should split our code to allow the file content retrieval to be mocked, by using a `io.ReadCloser` interface that you replace with a custom one in test context.

I have something like this in mind

```go
type TFCloudConfigReader struct {
  reader io.ReadCloser
}

func NewTFCloudConfigReader(reader io.ReadCloser) *TFCloudConfigReader {
  return &TFCloudConfigReader{reader}
}

func (r *TFCloudConfigReader) GetToken() (string, error) {
  // Build the config file path here or in an unexported func
  defer r.Close()
  var content []byte
  file.Read(&content)
  // Json stuff & cie
}

// In the real code
file:= os.Open(file)
reader := NewTFCloudConfigReader(file)
token := reader.GetToken()

// In tests
reader := backend.MockReaderMock() // We already have a mock for `ReadCloser`
reader.On("Close") // do nothing
reader.On("Read").Return() // Return your test content here
reader := NewTFCloudConfigReader(reader)
token := reader.GetToken()
```

Tell me if this is not clear we can do a quick pair session for this if you want.

Reply:
Maybe I misunderstood what you wrote at the end, but to be able to do `os.Open(file)` I need in the `real code` the config file path which means that at the beginning of the `GetToken` method we won't do any build of the path.

Secondly, I think you made a mistake inside the `GetToken` method where it should be `r.Read` or `r.reader.Read`. But even with that, I don't see how we could read the file with the`Read` method without the file's actual size.

What I'm up to atm:

```go
func (r *TFCloudConfigReader) GetToken() (string, error) {
	defer r.reader.Close()

	b, err := ioutil.ReadAll(r.reader)
	if err != nil {
		return "", err
	}

	var container Container
	if err := json.Unmarshal(b, &container); err != nil {
		return "", err
	}
	if container.Credentials.TerraformCloud.Token == "" {
		return "", errors.New("malformed JSON file: token not found")
	}
	return container.Credentials.TerraformCloud.Token, nil
}
```

And then I could do in the real code:

```go
tfConfigFile := getTerraformConfigFile()
file := os.Open(tfConfigFile)
reader := NewTFCloudConfigReader(file)
token = reader.GetToken()
```

Reply:
Yeah you are right we should build the path outside the `GetToken`